# Notebook 2 – Data Collection & Integration

Project:
Predicting NFL Player Value to Identify Contract Overpayment Risk Using Machine Learning

Student:
William Reed

Course:
DSE6311 Capstone

Date:
July 2026

## Objective

This notebook collects and combines player statistics, player information, and contract data for the project. The final wide receiver dataset will be used for EDA, feature engineering, and modeling.

In [2]:
import polars as pl

In [3]:
import nflreadpy as nfl

dir(nfl)

['__all__',
 '__builtins__',
 '__cached__',
 '__doc__',
 '__file__',
 '__loader__',
 '__name__',
 '__package__',
 '__path__',
 '__spec__',
 '__version__',
 'cache',
 'clear_cache',
 'config',
 'downloader',
 'get_current_season',
 'get_current_week',
 'load_combine',
 'load_contracts',
 'load_depth_charts',
 'load_draft_picks',
 'load_ff_opportunity',
 'load_ff_playerids',
 'load_ff_rankings',
 'load_ffverse',
 'load_ftn_charting',
 'load_injuries',
 'load_nextgen_stats',
 'load_officials',
 'load_participation',
 'load_pbp',
 'load_pfr_advstats',
 'load_player_stats',
 'load_players',
 'load_rosters',
 'load_rosters_weekly',
 'load_schedules',
 'load_snap_counts',
 'load_stats',
 'load_team_stats',
 'load_teams',
 'load_trades',
 'utils_date',
 'version']

In [4]:
seasons = list(range(2012, 2026))

player_stats = nfl.load_player_stats(
    seasons=seasons,
    summary_level="reg"
)

players = nfl.load_players()
contracts = nfl.load_contracts()

In [5]:
print("player Statistics:", player_stats.shape)
print("Players:", players.shape)
print("Contracts:", contracts.shape)

player Statistics: (26830, 148)
Players: (25052, 39)
Contracts: (51994, 26)


In [6]:
player_stats.columns

['player_id',
 'player_name',
 'player_display_name',
 'position',
 'position_group',
 'headshot_url',
 'season',
 'season_type',
 'recent_team',
 'games',
 'completions',
 'attempts',
 'passing_yards',
 'passing_tds',
 'passing_interceptions',
 'sacks_suffered',
 'sack_yards_lost',
 'sack_fumbles',
 'sack_fumbles_lost',
 'passing_air_yards',
 'passing_yards_after_catch',
 'passing_first_downs',
 'passing_epa',
 'passing_cpoe',
 'passing_2pt_conversions',
 'pacr',
 'passing_10',
 'passing_16',
 'passing_20',
 'passing_40',
 'carries',
 'rushing_yards',
 'rushing_tds',
 'rushing_fumbles',
 'rushing_fumbles_lost',
 'rushing_first_downs',
 'rushing_epa',
 'rushing_2pt_conversions',
 'rushing_10',
 'rushing_12',
 'rushing_20',
 'rushing_40',
 'receptions',
 'targets',
 'receiving_yards',
 'receiving_tds',
 'receiving_fumbles',
 'receiving_fumbles_lost',
 'receiving_air_yards',
 'receiving_yards_after_catch',
 'receiving_first_downs',
 'receiving_epa',
 'receiving_2pt_conversions',
 'receiv

In [7]:
players.columns

['gsis_id',
 'display_name',
 'common_first_name',
 'first_name',
 'last_name',
 'short_name',
 'football_name',
 'suffix',
 'esb_id',
 'nfl_id',
 'pfr_id',
 'pff_id',
 'otc_id',
 'espn_id',
 'smart_id',
 'birth_date',
 'position_group',
 'position',
 'ngs_position_group',
 'ngs_position',
 'height',
 'weight',
 'headshot',
 'college_name',
 'college_conference',
 'jersey_number',
 'rookie_season',
 'last_season',
 'latest_team',
 'status',
 'ngs_status',
 'ngs_status_short_description',
 'years_of_experience',
 'pff_position',
 'pff_status',
 'draft_year',
 'draft_round',
 'draft_pick',
 'draft_team']

In [8]:
contracts.columns

['player',
 'position',
 'team',
 'is_active',
 'year_signed',
 'years',
 'value',
 'apy',
 'guaranteed',
 'apy_cap_pct',
 'inflated_value',
 'inflated_apy',
 'inflated_guaranteed',
 'player_page',
 'otc_id',
 'gsis_id',
 'height',
 'weight',
 'college',
 'draft_year',
 'draft_round',
 'draft_overall',
 'draft_team',
 'date_of_birth',
 'season_history',
 'contract_history']

In [9]:
player_stats.head()

player_id,player_name,player_display_name,position,position_group,headshot_url,season,season_type,recent_team,games,completions,attempts,passing_yards,passing_tds,passing_interceptions,sacks_suffered,sack_yards_lost,sack_fumbles,sack_fumbles_lost,passing_air_yards,passing_yards_after_catch,passing_first_downs,passing_epa,passing_cpoe,passing_2pt_conversions,pacr,passing_10,passing_16,passing_20,passing_40,carries,rushing_yards,rushing_tds,rushing_fumbles,rushing_fumbles_lost,rushing_first_downs,rushing_epa,…,fg_missed_0_19,fg_missed_20_29,fg_missed_30_39,fg_missed_40_49,fg_missed_50_59,fg_missed_60_,fg_made_list,fg_missed_list,fg_blocked_list,fg_made_distance,fg_missed_distance,fg_blocked_distance,pat_made,pat_att,pat_missed,pat_blocked,pat_pct,gwfg_made,gwfg_att,gwfg_missed,gwfg_blocked,gwfg_distance_list,pt_att,pt_blocked,pt_long,pt_yards,pt_inside_20,pt_out_of_bounds,pt_downed,pt_touchback,pt_fair_caught,pt_returned,pt_return_yards,pt_return_tds,pt_net_yards,fantasy_points,fantasy_points_ppr
str,str,str,str,str,str,i32,str,str,i32,i32,i32,i32,i32,i32,i32,i32,i32,i32,i32,i32,i32,f64,f64,i32,f64,i32,i32,i32,i32,i32,i32,i32,i32,i32,i32,f64,…,i32,i32,i32,i32,i32,i32,str,str,str,i32,i32,i32,i32,i32,i32,i32,f64,i32,i32,i32,i32,str,i32,i32,i32,i32,i32,i32,i32,i32,i32,i32,i32,i32,i32,f64,f64
"""00-0000108""","""D.Akers""","""David Akers""","""K""","""SPEC""","""https://static.www.nfl.com/ima…",2012,"""REG""","""SF""",16,0,0,0,0,0,0,0,0,0,0,0,0,null,null,0,null,0,0,0,0,0,0,0,0,0,0,null,…,0,0,1,5,4,0,"""40;43;63;36;48;29;29;36;40;19;…","""55;40;43;52;41;50;51;39;44;40""","""43;33;21""",1022,455,97,44,44,0,0,1.0,0,0,0,0,null,0,0,null,0,0,0,0,0,0,0,0,0,0,0.0,0.0
"""00-0000551""","""B.Ayanbadejo""","""Brendon Ayanbadejo""","""ILB""","""LB""","""https://static.www.nfl.com/ima…",2012,"""REG""","""BAL""",13,0,0,0,0,0,0,0,0,0,0,0,0,null,null,0,null,0,0,0,0,0,0,0,0,0,0,null,…,0,0,0,0,0,0,null,null,null,0,0,0,0,0,0,0,null,0,0,0,0,null,0,0,null,0,0,0,0,0,0,0,0,0,0,0.0,0.0
"""00-0000585""","""C.Bailey""","""Champ Bailey""","""CB""","""DB""","""https://static.www.nfl.com/ima…",2012,"""REG""","""DEN""",16,0,0,0,0,0,0,0,0,0,0,0,0,null,null,0,null,0,0,0,0,0,0,0,0,0,0,null,…,0,0,0,0,0,0,null,null,null,0,0,0,0,0,0,0,null,0,0,0,0,null,0,0,null,0,0,0,0,0,0,0,0,0,0,0.0,0.0
"""00-0000741""","""R.Barber""","""Ronde Barber""","""FS""","""DB""","""https://static.www.nfl.com/ima…",2012,"""REG""","""TB""",16,0,0,0,0,0,0,0,0,0,0,0,0,null,null,0,null,0,0,0,0,0,0,0,0,0,0,null,…,0,0,0,0,0,0,null,null,null,0,0,0,0,0,0,0,null,0,0,0,0,null,0,0,null,0,0,0,0,0,0,0,0,0,0,0.0,0.0
"""00-0000865""","""C.Batch""","""Charlie Batch""","""QB""","""QB""","""https://static.www.nfl.com/ima…",2012,"""REG""","""PIT""",2,45,70,475,1,4,3,-12,0,0,663,257,22,-1.646032,1.449535,0,0.71644,23,8,5,1,0,0,0,0,0,0,null,…,0,0,0,0,0,0,null,null,null,0,0,0,0,0,0,0,null,0,0,0,0,null,0,0,null,0,0,0,0,0,0,0,0,0,0,15.0,15.0


In [10]:
player_stats.shape

(26830, 148)

In [11]:
print("=== PLAYER STATISTICS ===")
print(f"Rows, Columns: {player_stats.shape}")
print("\nData Types:")
print(player_stats.dtypes)
print("\nMissing Values:")
print(player_stats.null_count())

=== PLAYER STATISTICS ===
Rows, Columns: (26830, 148)

Data Types:
[String, String, String, String, String, String, Int32, String, String, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Float64, Float64, Int32, Float64, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Float64, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Float64, Int32, Int32, Int32, Int32, Int32, Float64, Float64, Float64, Float64, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Float64, Float64, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Float64, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, String, String, String, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Float64,

In [12]:
print("=== PLAYERS ===")
print(f"Rows, Columns: {players.shape}")
print("\nData Types:")
print(players.dtypes)
print("\nMissing Values:")
print(players.null_count())

=== PLAYERS ===
Rows, Columns: (25052, 39)

Data Types:
[String, String, String, String, String, String, String, String, String, String, String, String, String, String, String, String, String, String, String, String, Int32, Int32, String, String, String, String, Int32, Int32, String, String, String, String, Int32, String, String, Int32, Int32, Int32, String]

Missing Values:
shape: (1, 39)
┌─────────┬────────────┬───────────┬───────────┬───┬───────────┬───────────┬───────────┬───────────┐
│ gsis_id ┆ display_na ┆ common_fi ┆ first_nam ┆ … ┆ draft_yea ┆ draft_rou ┆ draft_pic ┆ draft_tea │
│ ---     ┆ me         ┆ rst_name  ┆ e         ┆   ┆ r         ┆ nd        ┆ k         ┆ m         │
│ u32     ┆ ---        ┆ ---       ┆ ---       ┆   ┆ ---       ┆ ---       ┆ ---       ┆ ---       │
│         ┆ u32        ┆ u32       ┆ u32       ┆   ┆ u32       ┆ u32       ┆ u32       ┆ u32       │
╞═════════╪════════════╪═══════════╪═══════════╪═══╪═══════════╪═══════════╪═══════════╪═══════════╡
│

In [13]:
print("=== CONTRACTS ===")
print(f"Rows, Columns: {contracts.shape}")
print("\nData Types:")
print(contracts.dtypes)
print("\nMissing Values:")
print(contracts.null_count())

=== CONTRACTS ===
Rows, Columns: (51994, 26)

Data Types:
[String, String, String, Boolean, Int32, Int32, Float64, Float64, Float64, Float64, Float64, Float64, Float64, String, Int32, String, String, String, String, Int32, Int32, Int32, String, String, List(Struct({'year': String, 'team': String, 'base_salary': Float64, 'prorated_bonus': Float64, 'option_bonus': Float64, 'roster_bonus': Float64, 'guaranteed_salary': Float64, 'cap_number': Float64, 'cap_percent': Float64, 'cash_paid': Float64, 'workout_bonus': Float64, 'per_game_roster_bonus': Float64, 'other_bonus': Float64})), List(Struct({'team': String, 'contract_type': String, 'status': String, 'year_signed': Int32, 'yrs': Int32, 'total': Float64, 'apy': Float64, 'guarantees': Float64, 'amount_earned': Float64, 'percent_earned': Float64, 'effective_apy': Float64}))]

Missing Values:
shape: (1, 26)
┌────────┬──────────┬──────┬───────────┬───┬────────────┬──────────────┬─────────────┬─────────────┐
│ player ┆ position ┆ team ┆ is_act

In [14]:
for i, col in enumerate(player_stats.columns, start=1):
    print(f"{i:3}. {col}")

  1. player_id
  2. player_name
  3. player_display_name
  4. position
  5. position_group
  6. headshot_url
  7. season
  8. season_type
  9. recent_team
 10. games
 11. completions
 12. attempts
 13. passing_yards
 14. passing_tds
 15. passing_interceptions
 16. sacks_suffered
 17. sack_yards_lost
 18. sack_fumbles
 19. sack_fumbles_lost
 20. passing_air_yards
 21. passing_yards_after_catch
 22. passing_first_downs
 23. passing_epa
 24. passing_cpoe
 25. passing_2pt_conversions
 26. pacr
 27. passing_10
 28. passing_16
 29. passing_20
 30. passing_40
 31. carries
 32. rushing_yards
 33. rushing_tds
 34. rushing_fumbles
 35. rushing_fumbles_lost
 36. rushing_first_downs
 37. rushing_epa
 38. rushing_2pt_conversions
 39. rushing_10
 40. rushing_12
 41. rushing_20
 42. rushing_40
 43. receptions
 44. targets
 45. receiving_yards
 46. receiving_tds
 47. receiving_fumbles
 48. receiving_fumbles_lost
 49. receiving_air_yards
 50. receiving_yards_after_catch
 51. receiving_first_downs
 52. 

In [15]:
for i, col in enumerate(players.columns, start=1):
    print(f"{i:3}. {col}")

  1. gsis_id
  2. display_name
  3. common_first_name
  4. first_name
  5. last_name
  6. short_name
  7. football_name
  8. suffix
  9. esb_id
 10. nfl_id
 11. pfr_id
 12. pff_id
 13. otc_id
 14. espn_id
 15. smart_id
 16. birth_date
 17. position_group
 18. position
 19. ngs_position_group
 20. ngs_position
 21. height
 22. weight
 23. headshot
 24. college_name
 25. college_conference
 26. jersey_number
 27. rookie_season
 28. last_season
 29. latest_team
 30. status
 31. ngs_status
 32. ngs_status_short_description
 33. years_of_experience
 34. pff_position
 35. pff_status
 36. draft_year
 37. draft_round
 38. draft_pick
 39. draft_team


In [16]:
for i, col in enumerate(contracts.columns, start=1):
    print(f"{i:3}. {col}")

  1. player
  2. position
  3. team
  4. is_active
  5. year_signed
  6. years
  7. value
  8. apy
  9. guaranteed
 10. apy_cap_pct
 11. inflated_value
 12. inflated_apy
 13. inflated_guaranteed
 14. player_page
 15. otc_id
 16. gsis_id
 17. height
 18. weight
 19. college
 20. draft_year
 21. draft_round
 22. draft_overall
 23. draft_team
 24. date_of_birth
 25. season_history
 26. contract_history


In [17]:
print("Player Stats Key:")
print(player_stats.select("player_id").head())

Player Stats Key:
shape: (5, 1)
┌────────────┐
│ player_id  │
│ ---        │
│ str        │
╞════════════╡
│ 00-0000108 │
│ 00-0000551 │
│ 00-0000585 │
│ 00-0000741 │
│ 00-0000865 │
└────────────┘


In [18]:
[c for c in player_stats.columns if "id" in c.lower()]

['player_id', 'pt_inside_20']

In [19]:
[c for c in players.columns if "id" in c.lower()]

['gsis_id',
 'esb_id',
 'nfl_id',
 'pfr_id',
 'pff_id',
 'otc_id',
 'espn_id',
 'smart_id']

In [20]:
[c for c in contracts.columns if "id" in c.lower()]

['otc_id', 'gsis_id']

In [21]:
players.select([
    "gsis_id",
    "display_name"
]).head(10)

gsis_id,display_name
str,str
"""00-0028830""","""Isaako Aaitui"""
"""00-0038389""","""Israel Abanikanda"""
"""00-0024644""","""Jon Abbate"""
"""ABB498348""","""Vince Abbott"""
"""00-0031021""","""Jared Abbrederis"""
"""00-0032860""","""Mehdi Abdesmad"""
"""00-0028564""","""Isa Abdul-Quddus"""
"""00-0032104""","""Ameer Abdullah"""
"""00-0023663""","""Hamza Abdullah"""


In [22]:
player_stats.select([
    "player_id",
    "player_name"
]).head(10)

player_id,player_name
str,str
"""00-0000108""","""D.Akers"""
"""00-0000551""","""B.Ayanbadejo"""
"""00-0000585""","""C.Bailey"""
"""00-0000741""","""R.Barber"""
"""00-0000865""","""C.Batch"""
"""00-0001263""","""M.Birk"""
"""00-0001820""","""K.Brooking"""
"""00-0004091""","""P.Dawson"""
"""00-0004541""","""D.Driver"""


In [23]:
players.filter(
    pl.col("display_name").str.contains("Aaron Rodgers", literal=False)
).select(["display_name", "gsis_id", "esb_id", "nfl_id"])

display_name,gsis_id,esb_id,nfl_id
str,str,str,str
"""Aaron Rodgers""","""00-0023459""","""ROD339293""","""29851"""


In [24]:
player_stats.filter(
    pl.col("player_name").str.contains("Aaron Rodgers", literal=False)
).select(["player_name", "player_id"])

player_name,player_id
str,str


In [25]:
nfl_master = (
    player_stats
    .join(
        players,
        left_on="player_id",
        right_on="gsis_id",
        how="left",
        suffix="_player"
    )
    .join(
        contracts,
        left_on="player_id",
        right_on="gsis_id",
        how="left",
        suffix="_contract"
    )
)

In [26]:
print("Player Stats Shape:", player_stats.shape)
print("Master Dataset Shape:", nfl_master.shape)
print("Unique Players:", nfl_master["player_id"].n_unique())

nfl_master.head()

Player Stats Shape: (26830, 148)
Master Dataset Shape: (167971, 211)
Unique Players: 7002


player_id,player_name,player_display_name,position,position_group,headshot_url,season,season_type,recent_team,games,completions,attempts,passing_yards,passing_tds,passing_interceptions,sacks_suffered,sack_yards_lost,sack_fumbles,sack_fumbles_lost,passing_air_yards,passing_yards_after_catch,passing_first_downs,passing_epa,passing_cpoe,passing_2pt_conversions,pacr,passing_10,passing_16,passing_20,passing_40,carries,rushing_yards,rushing_tds,rushing_fumbles,rushing_fumbles_lost,rushing_first_downs,rushing_epa,…,last_season,latest_team,status,ngs_status,ngs_status_short_description,years_of_experience,pff_position,pff_status,draft_year,draft_round,draft_pick,draft_team,player,position_contract,team,is_active,year_signed,years,value,apy,guaranteed,apy_cap_pct,inflated_value,inflated_apy,inflated_guaranteed,player_page,otc_id_contract,height_contract,weight_contract,college,draft_year_contract,draft_round_contract,draft_overall,draft_team_contract,date_of_birth,season_history,contract_history
str,str,str,str,str,str,i32,str,str,i32,i32,i32,i32,i32,i32,i32,i32,i32,i32,i32,i32,i32,f64,f64,i32,f64,i32,i32,i32,i32,i32,i32,i32,i32,i32,i32,f64,…,i32,str,str,str,str,i32,str,str,i32,i32,i32,str,str,str,str,bool,i32,i32,f64,f64,f64,f64,f64,f64,f64,str,i32,str,str,str,i32,i32,i32,str,str,list[struct[13]],list[struct[11]]
"""00-0000108""","""D.Akers""","""David Akers""","""K""","""SPEC""","""https://static.www.nfl.com/ima…",2012,"""REG""","""SF""",16,0,0,0,0,0,0,0,0,0,0,0,0,null,null,0,null,0,0,0,0,0,0,0,0,0,0,null,…,2013,"""DET""","""ACT""",null,null,16,"""K""",null,null,null,null,null,"""David Akers""","""K""","""Lions""",false,0,1,1.005,1.005,0.1,0.0,0.0,0.0,0.0,"""https://overthecap.com/player/…",1272,"""5'10""""","""200""","""Louisville""",1997,null,null,"""Panthers""","""December 9, 1974""","[{""2013"",""Lions"",0.94,0.065,null,null,null,0.62,0.005,1.005,null,null,null}, {""Total"",""Total"",0.94,0.065,null,null,null,0.62,null,1.005,null,null,null}]","[{null,null,null,null,null,null,null,null,null,null,null}, {null,null,null,null,null,null,null,null,null,null,null}]"
"""00-0000551""","""B.Ayanbadejo""","""Brendon Ayanbadejo""","""ILB""","""LB""","""https://static.www.nfl.com/ima…",2012,"""REG""","""BAL""",13,0,0,0,0,0,0,0,0,0,0,0,0,null,null,0,null,0,0,0,0,0,0,0,0,0,0,null,…,2012,"""BAL""","""ACT""",null,null,10,"""LB""",null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null
"""00-0000585""","""C.Bailey""","""Champ Bailey""","""CB""","""DB""","""https://static.www.nfl.com/ima…",2012,"""REG""","""DEN""",16,0,0,0,0,0,0,0,0,0,0,0,0,null,null,0,null,0,0,0,0,0,0,0,0,0,0,null,…,2013,"""DEN""","""ACT""",null,null,15,"""CB""",null,1999,1,7,"""WAS""","""Champ Bailey""","""CB""","""Broncos""",false,2004,7,63.032025,9.004575,6.0,0.112,235.601573,33.657368,22.426845,"""https://overthecap.com/player/…",4,"""6'0""""","""192""","""Georgia""",1999,1,7,"""Commanders""","""June 22, 1978""","[{""1999"",""Redskins"",0.592,1.04,null,0.0,0.0,1.632,0.028,5.792,0.0,null,null}, {""1999"",""Redskins"",0.592,1.04,null,0.0,0.0,1.632,0.028,5.792,0.0,null,null}, … {""Total"",""Total"",58.969,30.831,null,14.032025,18.0,107.632025,null,107.632025,3.8,null,null}]","[{""Redskins"",""Drafted"",""Expired"",1999,5,10.17,2.034,5.2,11.6,1.141,2.32}, {""Redskins"",""UFA"",""Traded"",2004,1,6.801,6.801,6.801,0.0,0.0,0.0}, … {""Saints"",""SFA"",""Terminated"",2014,2,3.75,1.875,0.5,0.0,0.0,0.0}]"
"""00-0000585""","""C.Bailey""","""Champ Bailey""","""CB""","""DB""","""https://static.www.nfl.com/ima…",2012,"""REG""","""DEN""",16,0,0,0,0,0,0,0,0,0,0,0,0,null,null,0,null,0,0,0,0,0,0,0,0,0,0,null,…,2013,"""DEN""","""ACT""",null,null,15,"""CB""",null,1999,1,7,"""WAS""","""Champ Bailey""","""CB""","""Broncos""",false,2011,4,43.0,10.75,10.5,0.089,107.593769,26.898442,26.272897,"""https://overthecap.com/player/…",4,"""6'0""""","""192""","""Georgia""",1999,1,7,"""Commanders""","""June 22, 1978"""

In [27]:
join_validation = nfl_master.select([
    pl.len().alias("total_rows"),
    pl.col("display_name").is_null().sum().alias("missing_player_matches"),
    pl.col("player").is_null().sum().alias("missing_contract_matches")
])

join_validation

total_rows,missing_player_matches,missing_contract_matches
u32,u32,u32
167971,14,471


## 7. Contract Duplication

The merged dataset contains more rows than the player statistics dataset. The increase is caused by multiple contract records for individual players. 

In [28]:
contract_counts = (
    contracts
    .group_by("gsis_id")
    .len()
    .sort("len", descending=True)
)

contract_counts.head(20)

gsis_id,len
str,u32
null,3535
"""00-0032463""",37
"""00-0035406""",36
"""00-0032891""",35
"""00-0032951""",34
…,…
"""00-0035924""",30
"""00-0034966""",29
"""00-0026300""",29


In [29]:
# Count missing GSIS IDs in the contracts table

null_gsis = contracts.filter(pl.col("gsis_id").is_null())

print(f"Contract records: {contracts.height:,}")
print(f"Missing gsis_id: {null_gsis.height:,}")
print(f"Percent missing: {100 * null_gsis.height / contracts.height:.2f}%")

Contract records: 51,994
Missing gsis_id: 3,535
Percent missing: 6.80%


In [30]:
# Remove records without a GSIS ID and examine duplicates again

contracts_valid = contracts.filter(pl.col("gsis_id").is_not_null())

contract_counts = (
    contracts_valid
    .group_by("gsis_id")
    .len()
    .sort("len", descending=True)
)

contract_counts.head(20)

gsis_id,len
str,u32
"""00-0032463""",37
"""00-0035406""",36
"""00-0032891""",35
"""00-0032951""",34
"""00-0034461""",34
…,…
"""00-0026300""",29
"""00-0034966""",29
"""00-0030294""",28


In [31]:
# Count contract records with missing gsis ID's

null_gsis = contracts.filter(
    pl.col("gsis_id").is_null()
)

print(f"Total contract records: {contracts.height:,}")
print(f"missing gsis_id: {null_gsis.height:,}")
print(
    f"Percent missing: "f"{100 * null_gsis.height / contracts.height:.2f}%"
)

Total contract records: 51,994
missing gsis_id: 3,535
Percent missing: 6.80%


In [32]:
# Retain contract records with a valid GSIS ID

contracts_valid = contracts.filter(
    pl.col("gsis_id").is_not_null()
)

# Count contract records for each identified player

contract_counts_valid = (
    contracts_valid
    .group_by("gsis_id")
    .len(name="contract_record_count")
    .sort("contract_record_count", descending=True)
)

contract_counts_valid.head(20)

gsis_id,contract_record_count
str,u32
"""00-0032463""",37
"""00-0035406""",36
"""00-0032891""",35
"""00-0032951""",34
"""00-0034461""",34
…,…
"""00-0026300""",29
"""00-0034966""",29
"""00-0034623""",28


In [33]:
# Display all columns in the contracts dataset

print(f"Number of columns: {len(contracts.columns)}")

for column in contracts.columns:
    print(column)

Number of columns: 26
player
position
team
is_active
year_signed
years
value
apy
guaranteed
apy_cap_pct
inflated_value
inflated_apy
inflated_guaranteed
player_page
otc_id
gsis_id
height
weight
college
draft_year
draft_round
draft_overall
draft_team
date_of_birth
season_history
contract_history


In [34]:
# Display every contract record for one player

sample_player = "00-0032463"

(
    contracts_valid
    .filter(pl.col("gsis_id") == sample_player)
    .sort("year_signed")
)

player,position,team,is_active,year_signed,years,value,apy,guaranteed,apy_cap_pct,inflated_value,inflated_apy,inflated_guaranteed,player_page,otc_id,gsis_id,height,weight,college,draft_year,draft_round,draft_overall,draft_team,date_of_birth,season_history,contract_history
str,str,str,bool,i32,i32,f64,f64,f64,f64,f64,f64,f64,str,i32,str,str,str,str,i32,i32,i32,str,str,list[struct[13]],list[struct[11]]
"""Kyle Peko""","""IDL""","""Broncos""",false,2016,2,0.99,0.495,0.0,0.003,1.920448,0.960224,0.0,"""https://overthecap.com/player/…",5126,"""00-0032463""","""6'1""""","""305""","""Oregon State""",2016,null,null,"""Broncos""","""July 23, 1993""","[{""2016"",""Broncos"",0.45,null,null,null,null,0.026471,0.0,0.026471,null,null,null}, {""2016"",""Broncos"",0.45,null,null,null,null,0.026471,0.0,0.026471,null,null,null}, … {""Total"",""Total"",6.3942,null,null,null,null,2.604706,null,2.92665,null,null,null}]","[{""Broncos"",""SFA"",""Terminated"",2016,2,0.99,0.495,0.0,0.026471,0.027,0.026471}, {""Broncos"",""SFA"",""Terminated"",2017,1,0.54,0.54,0.0,0.0,0.0,0.0}, … {""Chargers"",""Practice"",""Expired"",2025,1,0.315,0.315,0.0,0.07,0.222,0.07}]"
"""Kyle Peko""","""IDL""","""Broncos""",false,2017,2,1.17,0.585,0.0,0.004,2.110204,1.055102,0.0,"""https://overthecap.com/player/…",5126,"""00-0032463""","""6'1""""","""305""","""Oregon State""",2016,null,null,"""Broncos""","""July 23, 1993""","[{""2016"",""Broncos"",0.45,null,null,null,null,0.026471,0.0,0.026471,null,null,null}, {""2016"",""Broncos"",0.45,null,null,null,null,0.026471,0.0,0.026471,null,null,null}, … {""Total"",""Total"",6.3942,null,null,null,null,2.604706,null,2.92665,null,null,null}]","[{""Broncos"",""SFA"",""Terminated"",2016,2,0.99,0.495,0.0,0.026471,0.027,0.026471}, {""Broncos"",""SFA"",""Terminated"",2017,1,0.54,0.54,0.0,0.0,0.0,0.0}, … {""Chargers"",""Practice"",""Expired"",2025,1,0.315,0.315,0.0,0.07,0.222,0.07}]"
"""Kyle Peko""","""IDL""","""Broncos""",false,2017,1,0.54,0.54,0.0,0.003,0.97394,0.97394,0.0,"""https://overthecap.com/player/…",5126,"""00-0032463""","""6'1""""","""305""","""Oregon State""",2016,null,null,"""Broncos""","""July 23, 1993""","[{""2016"",""Broncos"",0.45,null,null,null,null,0.026471,0.0,0.026471,null,null,null}, {""2016"",""Broncos"",0.45,null,null,null,null,0.026471,0.0,0.026471,null,null,null}, … {""Total"",""Total"",6.3942,null,null,null,null,2.604706,null,2.92665,null,null,null}]","[{""Broncos"",""SFA"",""Terminated"",2016,2,0.99,0.495,0.0,0.026471,0.027,0.026471}, {""Broncos"",""SFA"",""Terminated"",2017,1,0.54,0.54,0.0,0.0,0.0,0.0}, … {""Chargers"",""Practice"",""Expired"",2025,1,0.315,0.315,0.0,0.07,0.222,0.07}]"
"""Kyle Peko""","""IDL""","""Broncos""",false,2017,1,0.54,0.54,0.0,0.003,0.97394,0.97394,0.0,"""https://overthecap.com/player/…",5126,"""00-0032463""","""6'1""""","""305""","""Oregon State""",2016,null,null,"""Broncos""","""July 23, 1993""","[{""2016"",""Broncos"",0.45,null,null,null,null,0.026471,0.0,0.026471,null,null,null}, {""2016"",""Broncos"",0.45,null,null,null,null,0.026471,0.0,0.026471,null,null,null}, … {""Total"",""Total"",6.3942,null,null,null,null,2.604706,null,2.92665,null,null,null}]","[{""Broncos"",""SFA"",""Terminated"",2016,2,0.99,0.495,0.0,0.026471,0.027,0.026471}, {""Broncos"",""SFA"",""Terminated"",2017,1,0.54,0.54,0.0,0.0,0.0,0.0}, … {""Chargers"",""Practice"",""Expired"",2025,1,0.315,0.315,0.0,0.07,0.222,0.07}]"
"""Kyle Peko""","""IDL""","""Broncos""",false,2017,1,0.1224,0.1224,0.0,0.001,0.22076,0.22076,0.0,"""https://overthecap.com/player/…",5126,"""00-0032463""","""6'1""""","""305""","""Oregon State""",2016,null,null,"""Broncos""","""July 23, 1993""","[{""2016"",""Broncos"",0.45,null,null,null,null,0.026471,0.0,0.026471,null,null,null}, {""2016"",""Broncos"",0.45,null,null,null,null,0.026471,0.0,0.026471,null,null,null}, … {""Total"",""Total"",6.3942,null,null,null,null,2.604706,null,2.92665,null,null,null}]","[{""Broncos"",""SFA"",""Terminated"",2016,2,0.99,0.495,0.0,0.026471,0.027,0.026471}, {""Broncos"",""

In [35]:
# Count exact duplicate rows in the contracts dataset

contracts_unique = contracts.unique()

duplicate_count = contracts.height - contracts_unique.height

print(f"Original rows: {contracts.height:,}")
print(f"Unique rows: {contracts_unique.height:,}")
print(f"Exact duplicate rows: {duplicate_count:,}")

Original rows: 51,994
Unique rows: 48,690
Exact duplicate rows: 3,304


### Current Findings

The contract data contain 4,190 rows without a `gsis_id`, so those records cannot be joined to the player statistics.

Players can also have many legitimate contract records over their careers. The dataset contains exact duplicate contract rows as well, so a direct contract merge would create repeated player season observations.

In [36]:
# Select one player and remove the nested "cols" column for easier comparison

sample_player = "00-0032463"

(
    contracts_valid
    .filter(pl.col("gsis_id") == sample_player)
    .sort(["year_signed", "team", "value"])
)

player,position,team,is_active,year_signed,years,value,apy,guaranteed,apy_cap_pct,inflated_value,inflated_apy,inflated_guaranteed,player_page,otc_id,gsis_id,height,weight,college,draft_year,draft_round,draft_overall,draft_team,date_of_birth,season_history,contract_history
str,str,str,bool,i32,i32,f64,f64,f64,f64,f64,f64,f64,str,i32,str,str,str,str,i32,i32,i32,str,str,list[struct[13]],list[struct[11]]
"""Kyle Peko""","""IDL""","""Broncos""",false,2016,2,0.99,0.495,0.0,0.003,1.920448,0.960224,0.0,"""https://overthecap.com/player/…",5126,"""00-0032463""","""6'1""""","""305""","""Oregon State""",2016,null,null,"""Broncos""","""July 23, 1993""","[{""2016"",""Broncos"",0.45,null,null,null,null,0.026471,0.0,0.026471,null,null,null}, {""2016"",""Broncos"",0.45,null,null,null,null,0.026471,0.0,0.026471,null,null,null}, … {""Total"",""Total"",6.3942,null,null,null,null,2.604706,null,2.92665,null,null,null}]","[{""Broncos"",""SFA"",""Terminated"",2016,2,0.99,0.495,0.0,0.026471,0.027,0.026471}, {""Broncos"",""SFA"",""Terminated"",2017,1,0.54,0.54,0.0,0.0,0.0,0.0}, … {""Chargers"",""Practice"",""Expired"",2025,1,0.315,0.315,0.0,0.07,0.222,0.07}]"
"""Kyle Peko""","""IDL""","""Broncos""",false,2017,1,0.1224,0.1224,0.0,0.001,0.22076,0.22076,0.0,"""https://overthecap.com/player/…",5126,"""00-0032463""","""6'1""""","""305""","""Oregon State""",2016,null,null,"""Broncos""","""July 23, 1993""","[{""2016"",""Broncos"",0.45,null,null,null,null,0.026471,0.0,0.026471,null,null,null}, {""2016"",""Broncos"",0.45,null,null,null,null,0.026471,0.0,0.026471,null,null,null}, … {""Total"",""Total"",6.3942,null,null,null,null,2.604706,null,2.92665,null,null,null}]","[{""Broncos"",""SFA"",""Terminated"",2016,2,0.99,0.495,0.0,0.026471,0.027,0.026471}, {""Broncos"",""SFA"",""Terminated"",2017,1,0.54,0.54,0.0,0.0,0.0,0.0}, … {""Chargers"",""Practice"",""Expired"",2025,1,0.315,0.315,0.0,0.07,0.222,0.07}]"
"""Kyle Peko""","""IDL""","""Broncos""",false,2017,1,0.1224,0.1224,0.0,0.001,0.22076,0.22076,0.0,"""https://overthecap.com/player/…",5126,"""00-0032463""","""6'1""""","""305""","""Oregon State""",2016,null,null,"""Broncos""","""July 23, 1993""","[{""2016"",""Broncos"",0.45,null,null,null,null,0.026471,0.0,0.026471,null,null,null}, {""2016"",""Broncos"",0.45,null,null,null,null,0.026471,0.0,0.026471,null,null,null}, … {""Total"",""Total"",6.3942,null,null,null,null,2.604706,null,2.92665,null,null,null}]","[{""Broncos"",""SFA"",""Terminated"",2016,2,0.99,0.495,0.0,0.026471,0.027,0.026471}, {""Broncos"",""SFA"",""Terminated"",2017,1,0.54,0.54,0.0,0.0,0.0,0.0}, … {""Chargers"",""Practice"",""Expired"",2025,1,0.315,0.315,0.0,0.07,0.222,0.07}]"
"""Kyle Peko""","""IDL""","""Broncos""",false,2017,1,0.54,0.54,0.0,0.003,0.97394,0.97394,0.0,"""https://overthecap.com/player/…",5126,"""00-0032463""","""6'1""""","""305""","""Oregon State""",2016,null,null,"""Broncos""","""July 23, 1993""","[{""2016"",""Broncos"",0.45,null,null,null,null,0.026471,0.0,0.026471,null,null,null}, {""2016"",""Broncos"",0.45,null,null,null,null,0.026471,0.0,0.026471,null,null,null}, … {""Total"",""Total"",6.3942,null,null,null,null,2.604706,null,2.92665,null,null,null}]","[{""Broncos"",""SFA"",""Terminated"",2016,2,0.99,0.495,0.0,0.026471,0.027,0.026471}, {""Broncos"",""SFA"",""Terminated"",2017,1,0.54,0.54,0.0,0.0,0.0,0.0}, … {""Chargers"",""Practice"",""Expired"",2025,1,0.315,0.315,0.0,0.07,0.222,0.07}]"
"""Kyle Peko""","""IDL""","""Broncos""",false,2017,1,0.54,0.54,0.0,0.003,0.97394,0.97394,0.0,"""https://overthecap.com/player/…",5126,"""00-0032463""","""6'1""""","""305""","""Oregon State""",2016,null,null,"""Broncos""","""July 23, 1993""","[{""2016"",""Broncos"",0.45,null,null,null,null,0.026471,0.0,0.026471,null,null,null}, {""2016"",""Broncos"",0.45,null,null,null,null,0.026471,0.0,0.026471,null,null,null}, … {""Total"",""Total"",6.3942,null,null,null,null,2.604706,null,2.92665,null,null,null}]","[{""Broncos"",""SFA"",""Terminated"",2016,2,0.99,0.495,0.0,0.026471,0.027,0.026471}, {""Broncos"","

### Findings Summary

The contracts dataset was examined to understand the increase in rows following integration with the player statistics dataset.

findings:

- 4,190 contract records (8.10%) contain missing `gsis_id` values and cannot be linked to player statistics.
- Many players legitimately possess multiple contract records because they signed multiple contracts throughout their careers.
- The contracts dataset also contains exact duplicate records. A comparison of individual player histories confirmed that several contracts are duplicated without any differences across the available fields.
- The increased row count after merging is attributable to both, one-to-many relationships and duplicate records present in the source contracts dataset.

No records were modified or removed yet. Data cleaning decisions will be deferred revisited during preprocessing stage after EDA.

In [37]:
player_season_master = (
    player_stats
    .join(
        players,
        left_on="player_id",
        right_on="gsis_id",
        how="left"
    )
)

print(player_season_master.shape)

(26830, 186)


In [38]:
wr_master = (
    player_season_master
    .filter(pl.col("position") == "WR")
)

print(f"Wide receiver records: {wr_master.height:,}")

wr_master.select("position").unique()

Wide receiver records: 3,151


position
str
"""WR"""


In [39]:
print(player_season_master.shape)
print(wr_master.shape)

(26830, 186)
(3151, 186)


In [40]:
(
    player_season_master
    .select("season")
    .unique()
    .sort("season")
)

season
i32
2012
2013
2014
2015
2016
…
2021
2022
2023


In [41]:
(
    player_season_master
    .group_by(["player_id", "season"])
    .len()
    .filter(pl.col("len") > 1)
)

player_id,season,len
str,i32,u32


## Summary

The player statistics and player metadata were successfully combined using the GSIS player identifier. The resulting wide receiver dataset contains one row per player and season.

A direct contract merge was not used in the final master dataset because players can have multiple contract records and the source also contains duplicate contract rows. Contract matching will be handled separately later in the project.

The final wr_master.csv dataset contains 3,151 wide receiver player season observations and 186 columns covering seasons 2012 through 2025.

In [42]:
from pathlib import Path

output_path = Path("../data/processed/wr_master.csv")
output_path.parent.mkdir(parents=True, exist_ok=True)

wr_master.write_csv(output_path)

print("Export complete!")
print(f"Location: {output_path.resolve()}")
print(f"Rows: {wr_master.height:,}")
print(f"Columns: {wr_master.width}")

Export complete!
Location: C:\Users\billf\Capstone-Project\data\processed\wr_master.csv
Rows: 3,151
Columns: 186


In [43]:
# Diagnose player-season duplicates in current WR master

duplicate_player_seasons = (
    wr_master
    .group_by(["player_id", "season"])
    .len()
    .filter(pl.col("len") > 1)
    .sort(["player_id", "season"])
)

print("WR rows:", wr_master.height)
print("Unique player-season keys:",
      wr_master.select(["player_id", "season"]).unique().height)
print("Duplicate player-season groups:", duplicate_player_seasons.height)

duplicate_player_seasons

WR rows: 3151
Unique player-season keys: 3151
Duplicate player-season groups: 0


player_id,season,len
str,i32,u32


In [57]:
# Inspect position fields for the current WR master

position_check = (
    wr_master
    .group_by(["position", "position_group"])
    .len()
    .sort("len", descending=True)
)

position_check

position,position_group,len
str,str,u32
"""WR""","""WR""",3151


In [44]:
# Inspect contract history for a sample player

sample_player = "00-0032463"

contracts.filter(
    pl.col("gsis_id") == sample_player
).sort("year_signed")

player,position,team,is_active,year_signed,years,value,apy,guaranteed,apy_cap_pct,inflated_value,inflated_apy,inflated_guaranteed,player_page,otc_id,gsis_id,height,weight,college,draft_year,draft_round,draft_overall,draft_team,date_of_birth,season_history,contract_history
str,str,str,bool,i32,i32,f64,f64,f64,f64,f64,f64,f64,str,i32,str,str,str,str,i32,i32,i32,str,str,list[struct[13]],list[struct[11]]
"""Kyle Peko""","""IDL""","""Broncos""",false,2016,2,0.99,0.495,0.0,0.003,1.920448,0.960224,0.0,"""https://overthecap.com/player/…",5126,"""00-0032463""","""6'1""""","""305""","""Oregon State""",2016,null,null,"""Broncos""","""July 23, 1993""","[{""2016"",""Broncos"",0.45,null,null,null,null,0.026471,0.0,0.026471,null,null,null}, {""2016"",""Broncos"",0.45,null,null,null,null,0.026471,0.0,0.026471,null,null,null}, … {""Total"",""Total"",6.3942,null,null,null,null,2.604706,null,2.92665,null,null,null}]","[{""Broncos"",""SFA"",""Terminated"",2016,2,0.99,0.495,0.0,0.026471,0.027,0.026471}, {""Broncos"",""SFA"",""Terminated"",2017,1,0.54,0.54,0.0,0.0,0.0,0.0}, … {""Chargers"",""Practice"",""Expired"",2025,1,0.315,0.315,0.0,0.07,0.222,0.07}]"
"""Kyle Peko""","""IDL""","""Broncos""",false,2017,2,1.17,0.585,0.0,0.004,2.110204,1.055102,0.0,"""https://overthecap.com/player/…",5126,"""00-0032463""","""6'1""""","""305""","""Oregon State""",2016,null,null,"""Broncos""","""July 23, 1993""","[{""2016"",""Broncos"",0.45,null,null,null,null,0.026471,0.0,0.026471,null,null,null}, {""2016"",""Broncos"",0.45,null,null,null,null,0.026471,0.0,0.026471,null,null,null}, … {""Total"",""Total"",6.3942,null,null,null,null,2.604706,null,2.92665,null,null,null}]","[{""Broncos"",""SFA"",""Terminated"",2016,2,0.99,0.495,0.0,0.026471,0.027,0.026471}, {""Broncos"",""SFA"",""Terminated"",2017,1,0.54,0.54,0.0,0.0,0.0,0.0}, … {""Chargers"",""Practice"",""Expired"",2025,1,0.315,0.315,0.0,0.07,0.222,0.07}]"
"""Kyle Peko""","""IDL""","""Broncos""",false,2017,1,0.54,0.54,0.0,0.003,0.97394,0.97394,0.0,"""https://overthecap.com/player/…",5126,"""00-0032463""","""6'1""""","""305""","""Oregon State""",2016,null,null,"""Broncos""","""July 23, 1993""","[{""2016"",""Broncos"",0.45,null,null,null,null,0.026471,0.0,0.026471,null,null,null}, {""2016"",""Broncos"",0.45,null,null,null,null,0.026471,0.0,0.026471,null,null,null}, … {""Total"",""Total"",6.3942,null,null,null,null,2.604706,null,2.92665,null,null,null}]","[{""Broncos"",""SFA"",""Terminated"",2016,2,0.99,0.495,0.0,0.026471,0.027,0.026471}, {""Broncos"",""SFA"",""Terminated"",2017,1,0.54,0.54,0.0,0.0,0.0,0.0}, … {""Chargers"",""Practice"",""Expired"",2025,1,0.315,0.315,0.0,0.07,0.222,0.07}]"
"""Kyle Peko""","""IDL""","""Broncos""",false,2017,1,0.54,0.54,0.0,0.003,0.97394,0.97394,0.0,"""https://overthecap.com/player/…",5126,"""00-0032463""","""6'1""""","""305""","""Oregon State""",2016,null,null,"""Broncos""","""July 23, 1993""","[{""2016"",""Broncos"",0.45,null,null,null,null,0.026471,0.0,0.026471,null,null,null}, {""2016"",""Broncos"",0.45,null,null,null,null,0.026471,0.0,0.026471,null,null,null}, … {""Total"",""Total"",6.3942,null,null,null,null,2.604706,null,2.92665,null,null,null}]","[{""Broncos"",""SFA"",""Terminated"",2016,2,0.99,0.495,0.0,0.026471,0.027,0.026471}, {""Broncos"",""SFA"",""Terminated"",2017,1,0.54,0.54,0.0,0.0,0.0,0.0}, … {""Chargers"",""Practice"",""Expired"",2025,1,0.315,0.315,0.0,0.07,0.222,0.07}]"
"""Kyle Peko""","""IDL""","""Broncos""",false,2017,1,0.1224,0.1224,0.0,0.001,0.22076,0.22076,0.0,"""https://overthecap.com/player/…",5126,"""00-0032463""","""6'1""""","""305""","""Oregon State""",2016,null,null,"""Broncos""","""July 23, 1993""","[{""2016"",""Broncos"",0.45,null,null,null,null,0.026471,0.0,0.026471,null,null,null}, {""2016"",""Broncos"",0.45,null,null,null,null,0.026471,0.0,0.026471,null,null,null}, … {""Total"",""Total"",6.3942,null,null,null,null,2.604706,null,2.92665,null,null,null}]","[{""Broncos"",""SFA"",""Terminated"",2016,2,0.99,0.495,0.0,0.026471,0.027,0.026471}, {""Broncos"",""

In [45]:
# Inspect duplicate contract records for wide receivers

wr_contracts = contracts.filter(
    pl.col("position") == "WR"
)

duplicate_summary = (
    wr_contracts
    .group_by(["gsis_id", "year_signed"])
    .agg(
        pl.len().alias("contract_rows"),
        pl.col("apy").n_unique().alias("unique_apy_values"),
        pl.col("team").n_unique().alias("unique_teams")
    )
    .filter(
        pl.col("contract_rows") > 1
    )
    .sort(
        "contract_rows",
        descending=True
    )
)

print("Wide receiver contract rows:", wr_contracts.height)
print(
    "Player and signing year combinations with multiple rows:",
    duplicate_summary.height
)

duplicate_summary.head(20)

Wide receiver contract rows: 7537
Player and signing year combinations with multiple rows: 1423


gsis_id,year_signed,contract_rows,unique_apy_values,unique_teams
str,i32,u32,u32,u32
null,2022,123,32,32
null,2021,107,28,30
null,2020,71,27,22
null,2019,46,17,23
null,2018,45,13,23
…,…,…,…,…
"""00-0034317""",2022,9,3,2
"""00-0032128""",2022,9,6,3
"""00-0035976""",2023,9,3,4


In [46]:
# Quantify usable WR contract records and exact duplicates

wr_contracts_valid = wr_contracts.filter(
    pl.col("gsis_id").is_not_null()
)

contract_core_columns = [
    "gsis_id",
    "year_signed",
    "team",
    "years",
    "value",
    "apy",
    "guaranteed",
    "apy_cap_pct",
    "inflated_value",
    "inflated_apy",
    "inflated_guaranteed",
]

wr_contracts_unique = (
    wr_contracts_valid
    .unique(
        subset=contract_core_columns,
        maintain_order=True
    )
)

print("All WR contract rows:", wr_contracts.height)
print(
    "WR rows with valid gsis_id:",
    wr_contracts_valid.height
)
print(
    "WR rows removed for missing gsis_id:",
    wr_contracts.height - wr_contracts_valid.height
)

print(
    "\nExact duplicate contract rows removed:",
    wr_contracts_valid.height - wr_contracts_unique.height
)

print(
    "Unique usable WR contract rows:",
    wr_contracts_unique.height
)

print(
    "Unique WR players with contract records:",
    wr_contracts_unique["gsis_id"].n_unique()
)

All WR contract rows: 7537
WR rows with valid gsis_id: 6912
WR rows removed for missing gsis_id: 625

Exact duplicate contract rows removed: 534
Unique usable WR contract rows: 6378
Unique WR players with contract records: 1626


In [47]:
# Recheck multiple contracts after removing missing IDs and exact duplicates

wr_contract_year_summary = (
    wr_contracts_unique
    .group_by(["gsis_id", "year_signed"])
    .agg(
        pl.len().alias("contract_rows"),
        pl.col("apy").n_unique().alias("unique_apy_values"),
        pl.col("team").n_unique().alias("unique_teams")
    )
    .filter(
        pl.col("contract_rows") > 1
    )
    .sort(
        "contract_rows",
        descending=True
    )
)

print(
    "Player and signing year combinations with multiple distinct contracts:",
    wr_contract_year_summary.height
)

print(
    "Maximum distinct contract rows for one player and signing year:",
    wr_contract_year_summary["contract_rows"].max()
)

wr_contract_year_summary.head(20)

Player and signing year combinations with multiple distinct contracts: 1355
Maximum distinct contract rows for one player and signing year: 7


gsis_id,year_signed,contract_rows,unique_apy_values,unique_teams
str,i32,u32,u32,u32
"""00-0034867""",2022,7,5,3
"""00-0032543""",2023,7,6,3
"""00-0032208""",2021,7,5,3
"""00-0032128""",2022,7,6,3
"""00-0035406""",2025,7,7,2
…,…,…,…,…
"""00-0035602""",2024,6,5,2
"""00-0032128""",2023,6,6,1
"""00-0038953""",2025,6,6,1


In [48]:
# Inspect one WR with multiple distinct contracts in the same year

sample_wr = "00-0032208"
sample_year = 2021

(
    wr_contracts_unique
    .filter(
        (pl.col("gsis_id") == sample_wr)
        & (pl.col("year_signed") == sample_year)
    )
    .select([
        "player",
        "position",
        "team",
        "year_signed",
        "years",
        "value",
        "apy",
        "guaranteed",
        "apy_cap_pct",
        "inflated_apy",
        "is_active",
    ])
    .sort("apy", descending=True)
)

player,position,team,year_signed,years,value,apy,guaranteed,apy_cap_pct,inflated_apy,is_active
str,str,str,i32,i32,f64,f64,f64,f64,f64,bool
"""Phillip Dorsett""","""WR""","""Jaguars""",2021,1,1.1275,1.1275,0.0,0.006,1.860838,false
"""Phillip Dorsett""","""WR""","""Texans""",2021,2,2.175,1.0875,0.3,0.006,1.794822,false
"""Phillip Dorsett""","""WR""","""Jaguars""",2021,1,0.293,0.293,0.0,0.002,0.48357,false
"""Phillip Dorsett""","""WR""","""Seahawks""",2021,1,0.292,0.292,0.0,0.002,0.48192,false
"""Phillip Dorsett""","""WR""","""Seahawks""",2021,1,0.252,0.252,0.0,0.001,0.415904,false
"""Phillip Dorsett""","""WR""","""Texans""",2021,1,0.252,0.252,0.0,0.001,0.415904,false
"""Phillip Dorsett""","""WR""","""Jaguars""",2021,1,0.252,0.252,0.0,0.001,0.415904,false


In [49]:
# Check contract ambiguity after including team

wr_contract_team_year_summary = (
    wr_contracts_unique
    .group_by([
        "gsis_id",
        "year_signed",
        "team"
    ])
    .agg(
        pl.len().alias("contract_rows"),
        pl.col("apy").n_unique().alias("unique_apy_values"),
        pl.col("years").n_unique().alias("unique_contract_lengths")
    )
    .filter(
        pl.col("contract_rows") > 1
    )
    .sort(
        "contract_rows",
        descending=True
    )
)

print(
    "Player, team, and signing year combinations with multiple contracts:",
    wr_contract_team_year_summary.height
)

print(
    "Maximum contracts within one player, team, and signing year:",
    wr_contract_team_year_summary["contract_rows"].max()
)

wr_contract_team_year_summary.head(20)

Player, team, and signing year combinations with multiple contracts: 960
Maximum contracts within one player, team, and signing year: 6


gsis_id,year_signed,team,contract_rows,unique_apy_values,unique_contract_lengths
str,i32,str,u32,u32,u32
"""00-0037132""",2022,"""Eagles""",6,6,3
"""00-0035406""",2025,"""Giants""",6,6,1
"""00-0040431""",2025,"""Giants""",6,6,3
"""00-0038891""",2025,"""Raiders""",6,6,2
"""00-0035876""",2022,"""Cardinals""",6,6,2
…,…,…,…,…,…
"""00-0033633""",2022,"""Bills""",5,5,2
"""00-0034867""",2022,"""49ers""",5,5,1
"""00-0036387""",2023,"""Patriots""",5,5,1


In [50]:
# Inspect detailed contract history for one WR

sample_wr_history = (
    wr_contracts_unique
    .filter(
        pl.col("gsis_id") == "00-0032208"
    )
    .select([
        "player",
        "gsis_id",
        "contract_history"
    ])
    .head(1)
)

print(sample_wr_history)

print("\nContract history contents:")
print(
    sample_wr_history[
        "contract_history"
    ][0]
)

shape: (1, 3)
┌─────────────────┬────────────┬─────────────────────────────────┐
│ player          ┆ gsis_id    ┆ contract_history                │
│ ---             ┆ ---        ┆ ---                             │
│ str             ┆ str        ┆ list[struct[11]]                │
╞═════════════════╪════════════╪═════════════════════════════════╡
│ Phillip Dorsett ┆ 00-0032208 ┆ [{"Patriots","Drafted","Renego… │
└─────────────────┴────────────┴─────────────────────────────────┘

Contract history contents:
shape: (325,)
Series: '' [struct[11]]
[
	{"Patriots","Drafted","Renegotiated",2015,4,8.13541,2.033853,6.591037,1.174582,0.144,1.174582}
	{"Colts","Drafted","Traded",2015,4,8.13541,2.033853,6.591037,5.416455,0.666,2.708228}
	{"Patriots","Other","Expired",2018,1,1.394373,1.394373,0.15,1.394373,1.0,1.394373}
	{"Patriots","UFA","Expired",2019,1,2.6,2.6,0.5,2.6,1.0,2.6}
	{"Seahawks","UFA","Expired",2020,1,1.0475,1.0475,0.1375,1.0475,1.0,1.0475}
	…
	{"Falcons","SFA","Terminated",2025,1,1.25

In [51]:
# Expand WR contract history into one row per contract event

wr_contract_history = (
    wr_contracts_unique
    .select([
        "gsis_id",
        "player",
        "position",
        "contract_history"
    ])
    .explode("contract_history")
    .unnest("contract_history")
)

print("Expanded WR contract history rows:", wr_contract_history.height)
print("\nColumns:")
print(wr_contract_history.columns)

wr_contract_history.head(10)

Expanded WR contract history rows: 404078

Columns:
['gsis_id', 'player', 'position', 'team', 'contract_type', 'status', 'year_signed', 'yrs', 'total', 'apy', 'guarantees', 'amount_earned', 'percent_earned', 'effective_apy']


C:\Users\billf\AppData\Local\Temp\ipykernel_33056\3677407991.py:11: DeprecationWarning: In Polars 2.0, the default behavior for `empty_as_null` will change to `False`. To keep the current behavior, explicitly set `empty_as_null=True`.
  .explode("contract_history")


gsis_id,player,position,team,contract_type,status,year_signed,yrs,total,apy,guarantees,amount_earned,percent_earned,effective_apy
str,str,str,str,str,str,i32,i32,f64,f64,f64,f64,f64,f64
"""00-0036900""","""Ja'Marr Chase""","""WR""","""Bengals""","""Drafted""","""Expired""",2021,4,30.819642,7.704911,30.819642,30.819642,1.0,7.704911
"""00-0036900""","""Ja'Marr Chase""","""WR""","""Bengals""","""Extension""","""Active""",2025,4,161.0,40.25,73.9,40.662843,0.253,40.662843
"""00-0036900""","""Ja'Marr Chase""","""WR""","""Bengals""","""Drafted""","""Expired""",2021,4,30.819642,7.704911,30.819642,30.819642,1.0,7.704911
"""00-0036900""","""Ja'Marr Chase""","""WR""","""Bengals""","""Extension""","""Active""",2025,4,161.0,40.25,73.9,40.662843,0.253,40.662843
"""00-0036900""","""Ja'Marr Chase""","""WR""","""Bengals""","""Drafted""","""Expired""",2021,4,30.819642,7.704911,30.819642,30.819642,1.0,7.704911
"""00-0036900""","""Ja'Marr Chase""","""WR""","""Bengals""","""Extension""","""Active""",2025,4,161.0,40.25,73.9,40.662843,0.253,40.662843
"""00-0036900""","""Ja'Marr Chase""","""WR""","""Bengals""","""Drafted""","""Expired""",2021,4,30.819642,7.704911,30.819642,30.819642,1.0,7.704911
"""00-0036900""","""Ja'Marr Chase""","""WR""","""Bengals""","""Extension""","""Active""",2025,4,161.0,40.25,73.9,40.662843,0.253,40.662843
"""00-0036900""","""Ja'Marr Chase""","""WR""","""Bengals""","""Drafted""","""Expired""",2021,4,30.819642,7.704911,30.819642,30.819642,1.0,7.704911


In [52]:
# Expand one contract history per WR into contract events

wr_player_histories = (
    wr_contracts_valid
    .select([
        "gsis_id",
        "player",
        "position",
        "contract_history"
    ])
    .unique(
        subset=["gsis_id"],
        maintain_order=True
    )
)

wr_contract_history = (
    wr_player_histories
    .explode("contract_history")
    .unnest("contract_history")
    .unique(
        subset=[
            "gsis_id",
            "team",
            "contract_type",
            "status",
            "year_signed",
            "yrs",
            "total",
            "apy",
            "guarantees",
            "amount_earned",
            "percent_earned",
            "effective_apy"
        ],
        maintain_order=True
    )
)

print(
    "Unique WR players:",
    wr_player_histories.height
)

print(
    "Unique WR contract events:",
    wr_contract_history.height
)

wr_contract_history.head(10)

Unique WR players: 1626
Unique WR contract events: 7122


C:\Users\billf\AppData\Local\Temp\ipykernel_33056\4193014234.py:19: DeprecationWarning: In Polars 2.0, the default behavior for `empty_as_null` will change to `False`. To keep the current behavior, explicitly set `empty_as_null=True`.
  .explode("contract_history")


gsis_id,player,position,team,contract_type,status,year_signed,yrs,total,apy,guarantees,amount_earned,percent_earned,effective_apy
str,str,str,str,str,str,i32,i32,f64,f64,f64,f64,f64,f64
"""00-0036900""","""Ja'Marr Chase""","""WR""","""Bengals""","""Drafted""","""Expired""",2021,4,30.819642,7.704911,30.819642,30.819642,1.0,7.704911
"""00-0036900""","""Ja'Marr Chase""","""WR""","""Bengals""","""Extension""","""Active""",2025,4,161.0,40.25,73.9,40.662843,0.253,40.662843
"""00-0033040""","""Tyreek Hill""","""WR""","""Chiefs""","""Drafted""","""Renegotiated""",2016,4,2.58622,0.646555,0.21872,1.806256,0.698,0.602085
"""00-0033040""","""Tyreek Hill""","""WR""","""Chiefs""","""Extension""","""Traded""",2019,3,54.0,18.0,18.34,36.02,0.667,12.006667
"""00-0033040""","""Tyreek Hill""","""WR""","""Dolphins""","""Extension""","""Renegotiated""",2022,4,120.0,30.0,52.535,53.235,0.444,26.6175
"""00-0033040""","""Tyreek Hill""","""WR""","""Dolphins""","""Extension""","""Terminated""",2024,3,90.0,30.0,54.0,52.623529,0.585,26.311765
"""00-0038543""","""Jaxon Smith-Njigba""","""WR""","""Seahawks""","""Drafted""","""Extended""",2023,4,14.417308,3.604327,14.417308,11.701312,0.812,3.900437
"""00-0038543""","""Jaxon Smith-Njigba""","""WR""","""Seahawks""","""Extension""","""Active""",2026,4,168.6,42.15,69.130996,35.0,0.208,0.0
"""00-0030564""","""DeAndre Hopkins""","""WR""","""Texans""","""Drafted""","""Renegotiated""",2013,4,7.6267,1.906675,6.181696,7.6267,1.0,1.906675


In [53]:
# Review WR contract event types and statuses

contract_type_counts = (
    wr_contract_history
    .group_by("contract_type")
    .agg(
        pl.len().alias("events")
    )
    .sort(
        "events",
        descending=True
    )
)

status_counts = (
    wr_contract_history
    .group_by("status")
    .agg(
        pl.len().alias("events")
    )
    .sort(
        "events",
        descending=True
    )
)

print("Contract types:")
print(contract_type_counts)

print("\nContract statuses:")
print(status_counts)

Contract types:
shape: (12, 2)
┌───────────────┬────────┐
│ contract_type ┆ events │
│ ---           ┆ ---    │
│ str           ┆ u32    │
╞═══════════════╪════════╡
│ SFA           ┆ 2269   │
│ Practice      ┆ 2230   │
│ UDFA          ┆ 704    │
│ Drafted       ┆ 666    │
│ UFA           ┆ 511    │
│ …             ┆ …      │
│ ERFA          ┆ 150    │
│ RFA           ┆ 84     │
│ Other         ┆ 76     │
│ Franchise     ┆ 25     │
│               ┆ 1      │
└───────────────┴────────┘

Contract statuses:
shape: (13, 2)
┌───────────────────┬────────┐
│ status            ┆ events │
│ ---               ┆ ---    │
│ str               ┆ u32    │
╞═══════════════════╪════════╡
│ Terminated        ┆ 3313   │
│ Expired           ┆ 1770   │
│ Renegotiated      ┆ 548    │
│ Elevated          ┆ 545    │
│ Active            ┆ 400    │
│ …                 ┆ …      │
│ Extended          ┆ 63     │
│ Reserve           ┆ 22     │
│ June 1 Terminated ┆ 5      │
│ Deceased          ┆ 1      │
│ Inactive

In [54]:
status_counts.to_dicts()

[{'status': 'Terminated', 'events': 3313},
 {'status': 'Expired', 'events': 1770},
 {'status': 'Renegotiated', 'events': 548},
 {'status': 'Elevated', 'events': 545},
 {'status': 'Active', 'events': 400},
 {'status': None, 'events': 242},
 {'status': 'Traded', 'events': 139},
 {'status': 'Claimed', 'events': 73},
 {'status': 'Extended', 'events': 63},
 {'status': 'Reserve', 'events': 22},
 {'status': 'June 1 Terminated', 'events': 5},
 {'status': 'Deceased', 'events': 1},
 {'status': 'Inactive', 'events': 1}]

In [55]:
# Review contract history coverage by signing year

contract_year_coverage = (
    wr_contract_history
    .group_by("year_signed")
    .agg(
        pl.len().alias("contract_events"),
        pl.col("gsis_id").n_unique().alias("unique_players")
    )
    .sort("year_signed")
)

print(contract_year_coverage)

shape: (33, 3)
┌─────────────┬─────────────────┬────────────────┐
│ year_signed ┆ contract_events ┆ unique_players │
│ ---         ┆ ---             ┆ ---            │
│ i32         ┆ u32             ┆ u32            │
╞═════════════╪═════════════════╪════════════════╡
│ null        ┆ 242             ┆ 242            │
│ 0           ┆ 1               ┆ 1              │
│ 1994        ┆ 2               ┆ 1              │
│ 1995        ┆ 1               ┆ 1              │
│ 1996        ┆ 1               ┆ 1              │
│ …           ┆ …               ┆ …              │
│ 2022        ┆ 671             ┆ 305            │
│ 2023        ┆ 663             ┆ 331            │
│ 2024        ┆ 687             ┆ 340            │
│ 2025        ┆ 692             ┆ 349            │
│ 2026        ┆ 330             ┆ 310            │
└─────────────┴─────────────────┴────────────────┘


In [56]:
# Focus contract coverage on the project analysis period

project_contract_coverage = (
    wr_contract_history
    .filter(
        pl.col("year_signed").is_between(2012, 2024)
    )
    .group_by("year_signed")
    .agg(
        pl.len().alias("contract_events"),
        pl.col("gsis_id").n_unique().alias("unique_players"),
        pl.col("apy").median().alias("median_apy"),
        pl.col("apy").max().alias("max_apy")
    )
    .sort("year_signed")
)

print(project_contract_coverage)

shape: (13, 5)
┌─────────────┬─────────────────┬────────────────┬────────────┬───────────┐
│ year_signed ┆ contract_events ┆ unique_players ┆ median_apy ┆ max_apy   │
│ ---         ┆ ---             ┆ ---            ┆ ---        ┆ ---       │
│ i32         ┆ u32             ┆ u32            ┆ f64        ┆ f64       │
╞═════════════╪═════════════════╪════════════════╪════════════╪═══════════╡
│ 2012        ┆ 112             ┆ 73             ┆ 0.627522   ┆ 16.207143 │
│ 2013        ┆ 110             ┆ 72             ┆ 0.5638815  ┆ 12.8435   │
│ 2014        ┆ 166             ┆ 120            ┆ 0.606948   ┆ 10.0      │
│ 2015        ┆ 163             ┆ 125            ┆ 0.690788   ┆ 15.0      │
│ 2016        ┆ 205             ┆ 157            ┆ 0.621353   ┆ 14.599    │
│ …           ┆ …               ┆ …              ┆ …          ┆ …         │
│ 2020        ┆ 555             ┆ 286            ┆ 0.585      ┆ 27.25     │
│ 2021        ┆ 596             ┆ 298            ┆ 0.68075    ┆ 18.0     